### PR TITLE
Give TestWorkingBeatmap a reference clock

### PR DESCRIPTION
**Pull request by peppy**
Previously, `TestWorkingBeatmap` using the default implementation of `TrackVirtual` with an underlying `Stopwatch` could potentially run far ahead of the underlying tests (in the case the tests were running non-realtime). 

This change gives the option of tying `TestWorkingBeatmap` to an underlying clock, resolving this abnormality.

### CODE REVIEW

File: osu.Game/Tests/Beatmaps/TestWorkingBeatmap.cs [hunk 1]
```diff
@@ -1,29 +1,106 @@
...
+
+            public override bool IsRunning => running;
+
+            public override double CurrentTime => running ? clock.CurrentTime : 0;
+
+            protected override void UpdateState()
+            {
+                base.UpdateState();
+
+                clock.CurrentTime = Math.Min(referenceClock.CurrentTime + offset, Length);
```
Comment (smoogipoo) [inline osu.Game/Tests/Beatmaps/TestWorkingBeatmap.cs]: Should probably respect the `running` state.

**Review by smoogipoo — DISMISSED** [review]

File: osu.Game/Tests/Beatmaps/TestWorkingBeatmap.cs [hunk 2]
```diff
@@ -82,18 +86,41 @@ public override void Reset()
...
+                    double refTime = referenceClock.CurrentTime;
+
+                    if (!lastReferenceTime.HasValue)
+                    {
+                        // if the clock just started running, the current value should be transferred to the offset
+                        // (to zero the progression of time).
+                        offset -= refTime;
+                    }
+
+                    lastReferenceTime = referenceClock.CurrentTime;
```
Comment (smoogipoo) [inline osu.Game/Tests/Beatmaps/TestWorkingBeatmap.cs]:
```suggestion
                    lastReferenceTime = refTime;
```

File: osu.Game/Tests/Beatmaps/TestWorkingBeatmap.cs [hunk 3]
```diff
@@ -50,28 +50,32 @@ public TestWorkingBeatmap(IBeatmap beatmap, IFrameBasedClock referenceClock = nu
         public class TrackVirtualManual : Track
         {
             private readonly IFrameBasedClock referenceClock;
-            private readonly ManualClock clock;
+
+            private readonly ManualClock clock = new ManualClock();
 
             private bool running;
```
Comment (smoogipoo) [inline osu.Game/Tests/Beatmaps/TestWorkingBeatmap.cs]: Intended for this to be false by default?

**Review by smoogipoo — DISMISSED** [review]

File: osu.Game/Tests/Beatmaps/TestWorkingBeatmap.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Tests/Beatmaps/TestWorkingBeatmap.cs]: seems correct, referencing `TrackVirtual`

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**